src/storage-client/src/source/persist_source.rs

```diff
@@ -134,465 +126,149 @@ where
-                            }
-                        }
-                    }
-                    Some(Err::<_, ExternalError>(e)) => {
-                        panic!("unexpected error from persist {e}")
-                    }
-                    // We never expect any further output from
-                    // `pinned_stream`, so propagate that information
-                    // downstream.
-                    None => {
-                        cap_set.downgrade(&[]);
-                        return;
-                    }
-                }
-            }
-
-            // We've exhausted our budget, listen for updates to the
-            // flow_control input's frontier until we free up new budget.
-            // Progress ChangeBatches are consumed even if you don't interact
-            // with the handle, so even if we never make it to this block (e.g.
-            // budget of usize::MAX), because the stream has no data, we don't
-            // cause unbounded buffering in timely.
-            while inflight_bytes >= flow_control_max_inflight_bytes {
-                // Get an upper bound until which we should produce data
-                let flow_control_upper = match flow_control_input.next().await {
-                    Some(Event::Progress(frontier)) => frontier,
-                    Some(Event::Data(_, _)) => unreachable!("flow_control_input should not contain data"),
-                    None => Antichain::new(),
-                };
-
-                let retired_parts = inflight_parts.drain_filter_swapping(|(upper, _size)| {
-                    PartialOrder::less_equal(&*upper, &flow_control_upper)
-                });
-
-                for (_upper, size_in_bytes) in retired_parts {
-                    inflight_bytes -= size_in_bytes;
-                    trace!(
-                        "shard {} returning {} bytes. total: {}. batch frontier {:?} less_equal to {:?}",
-                        data_shard,
-                        size_in_bytes,
-                        inflight_bytes,
-                        _upper,
-                        flow_control_upper,
-                    );
-                }
-            }
-        }
-    });
+    let rows = decode_and_mfp(&fetched, &name, until, map_filter_project, yield_fn);
+    (rows, token)
+}
 
-    let mut fetcher_builder = AsyncOperatorBuilder::new(
-        format!("persist_source {}: part fetcher", source_id),
-        scope.clone(),
-    );
+pub fn decode_and_mfp<G, YFn>(
+    fetched: &Stream<G, FetchedPart<SourceData, (), Timestamp, Diff>>,
+    name: &str,
+    until: Antichain<Timestamp>,
+    mut map_filter_project: Option<&mut MfpPlan>,
+    yield_fn: YFn,
+) -> Stream<G, (Result<Row, DataflowError>, Timestamp, Diff)>
+where
+    G: Scope<Timestamp = mz_repr::Timestamp>,
+    YFn: Fn(Instant, usize) -> bool + 'static,
+{
+    let mut builder =
+        AsyncOperatorBuilder::new(format!("decode_and_mfp({})", name), fetched.scope());
```

Comment:
should be sth like `persist_source::decode_and_mfp`, or `persist_source_decode_and_mfp`

Reply:
done!